pkg/expansion/graph.go

```diff
@@ -0,0 +1,116 @@
+package expansion
+
+import (
+	"fmt"
+
+	expansionunversioned "github.com/open-policy-agent/gatekeeper/apis/expansion/unversioned"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type cycleDetector interface {
+	addTemplate(*expansionunversioned.ExpansionTemplate) error
+	removeTemplate(*expansionunversioned.ExpansionTemplate)
+}
+
+// gvkGraph represents a directed acyclic graph of GVKs. The terminal int value
+// represents how many edges exists between two GVKs.
+type gvkGraph map[schema.GroupVersionKind]map[schema.GroupVersionKind]int
+
+// addTemplate tries to upsert the template to the graph, returning an error if
+// a cycle is created. No edges are added if a cycle is encountered.
+func (g gvkGraph) addTemplate(template *expansionunversioned.ExpansionTemplate) error {
+	y := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+	xs := applyToGVKs(template)
+	for _, x := range xs {
+		if g.checkCycle(x, y) {
+			return fmt.Errorf("template forms expansion cycle starting with %v to %v", x, y)
+		}
+	}
+
+	for _, x := range xs {
+		g.addEdge(x, y)
+	}
+	return nil
+}
+
+func (g gvkGraph) removeTemplate(template *expansionunversioned.ExpansionTemplate) {
+	y := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+	xs := applyToGVKs(template)
+	for _, x := range xs {
+		g.removeEdge(x, y)
+	}
+}
+
+func (g gvkGraph) addEdge(x, y schema.GroupVersionKind) {
+	if _, exists := g[x]; !exists {
+		g[x] = make(map[schema.GroupVersionKind]int)
+	}
+	g[x][y]++
+}
+
+func (g gvkGraph) removeEdge(x, y schema.GroupVersionKind) {
+	if _, exists := g[x]; !exists {
+		return
+	}
+	if _, exists := g[x][y]; !exists {
+		return
+	}
+
+	g[x][y]--
+	// Prune empty edges
+	if g[x][y] == 0 {
+		delete(g[x], y)
+		if len(g[x]) == 0 {
+			delete(g, x)
+		}
+	}
+}
+
+// checkCycle returns true if adding an edge from `x` to `y` creates a cycle.
+func (g gvkGraph) checkCycle(x, y schema.GroupVersionKind) bool {
```

Comment:
if our goal is to have a tool that does cycle detection, could we leverage some package off the shelf? unless there's a strong reason to have a bespoke implementation of a graph in this case

briefly looked at https://github.com/dominikbraun/graph to see if it'd work here and I think it would! We could still have our interface and have an implementor just delegate to the underlying lib.

wdyt?

Reply:
One unique requirement is we can have multiple edges between the same node. Theoretically, you could define two expansion template which have the same `applyTo` and `generatedGVK`, but different `source` (e.g. maybe one source is `spec.foo`, the other `spec.bar`). 

It don't think the library supports that. We could maybe hack it using a weighed graph maybe but at that point it's perhaps not worth it 🤷 

I suspect with the changes needed to address Max's race-condition the requirements will diverge even more, because we will need to keep track of cyclic edges somehow without throwing them away.